central/processlisteningonport/README.md

```diff
@@ -79,3 +79,9 @@ Internal API:
 
 * ProcessListeningOnPort -- counter, represents how many PLOP objects Sensor
   has sent to the Central.
+
+Public API:
+
+* Central holds all listening endpoints for a deployment in memory when the
```

Comment:
Do we really want this level of memory consumption to simply display PLOP?  How is PLOP presented to the user such that this is necessary? Could we not use the database to sort?

I do not know enough about PLOP to have these answers, but these are the question that first come to mind that I think would help me understand.  Though admittedly this memory consumption gives me some pause.

Reply:
The first thing I will say is that sorting the listening endpoints does not seem to lead to significantly more memory consumption, so merging this PR will not change the memory consumption.

I don't have numbers on the typical number of listening endpoints per deployment. I think that I should have that. I think 200,000 listening endpoints in one single deployment is really high.

Lastly I might need to double check this. This was with running 10 queries. There may have been overlap between the calls to the API and  it is possible that the memory consumption was only ~200 Mi not 2 Gi.

Reply:
In the future pagination could reduce the memory consumption.

Reply:
Perhaps it would be better to look at doing that instead.

Reply:
Those things are not mutually exclusive. Except that if pagination is done, it would be good to move the sort into the query. That would require adding extra columns. The sorting could be kept as is with pagination, but then the sorting would be just within each page, rather than overall.

I did a new test where only one API call is made and I measure the memory usage. ~200,000 listening endpoints used 1 Gi, whereas when 10 API calls were made 2 Gi was the peak memory usage.

Reply:
I think we should take a step back here and think more broadly.   Define the problem we are going after and quantify its impact.  I'm not certain what the actual need is here which makes me uncomfortable with the potentially high memory consumption.  

Furthermore, adding this sort now and then maybe pushing sort to the DB when we add pagination later, seems like wasted effort.  

In some of the performance research that has been going on, we have seen some inefficiencies in PLOP we are trying to quantify.  That also makes me reluctant to add a resource intensive sort.

Reply:
With previous performance testing we had measured CPU and memory usage when adding listening endpoints, and we had measured various timings related to calling the API, but not how much memory is used by central when the API is called. In this example ~200,000 listening endpoints used ~1 Gi. That is about 5 Ki per listening endpoint if I did my math correctly. That seems way more than I would expect. According to K6 about 150 MB where returned from the API. There seems to be a huge discrepancy there.  I think the issue might be that it stored the serialized listening endpoint, the serialized process indicator, and then a PLOP object for each row. I don't think all of the data in both of the serialized columns is needed. I think there are some improvements we can make in the memory usage. Part of ProcessListeningOnPort is Signal, which is a pointer to a ProcessSignal. In the test with just one API call there were ~200,000 listening endpoints and only ~25,000 process indicators. In this case we would also have ~200,000 instances of ProcessSignal, whereas we only need ~25,000. Perhaps in most cases the ratio would be closer to 1:1. We could have a map, where the key is a process indicator id and the value is a pointer to a ProcessSignal. This might give us some significant memory savings in this case.

I also have some ideas on how to improve the speed of the API, but I would like to do some profiling to find out where the time is being taken before I propose changes.

Reply:
There are certainly gaps in my knowledge. I think @erthalion said something about the results of the postgres query being streamed. How much of the query result is held in memory by central at a time? Is that something we can tune?

Perhaps we don't need to save the results to `var plops []*storage.ProcessListeningOnPort`. Perhaps the results can be streamed as well. I don't know.

Reply:
The motivation for this work is that the listening endpoints are displayed in somewhat of a random order in the UI and also in the API. I guess the order in the API response is the order in which the listening endpoints were inserted into the database. The order in the UI is the same as the API and there is no option to change the order in the UI. The random ordering did bother me a bit when just had the API. I first liked the look of the UI, but every time that I looked at the API I liked it less and less, because of the random ordering. I think that customers don't like the random ordering either. I first went to the UI team to suggest to them to sort the listening endpoints. They suggested that I change the ordering in the API, since they just use the ordering from the API.

I still don't think that the resource consumption of sorting is significant. I don't think think it increases the peak memory usage. It just extends the time during which the memory usage is high by a small amount. I respect your concern for the resource usage.

Reply:
My 2c for this discussion:

* The PR itself has little to do with the resource consumption problem, it existed before that. From this point of view the issue could be addressed independently.
* The problem we discuss here seems to be common throughout the project. As far as I can tell (pls correct me if I'm wrong) there is a plenty of API calls not using pagination, or doing it with`Limit: math.MaxInt32`. Having this in mind, it's probably better to address the issue in a holistic way, since every such unbounded calls could be an issue.
* If I understand correctly, the current pagination implementation seems to use OFFSET, which is equivalent to a seq scan and poses a problem on it's own:
```
paginationSB.WriteString(fmt.Sprintf(" OFFSET %d", p.Offset))
```
* The best way of solving the issue of requesting an unbounded data is to use pagination. Any kind of streaming is possible, but only a work around (it's designed for those cases where pulling large chunk of data is necessary).